File: problem1.go
// Time Complexity : O(2^N)
// Space Complexity : O(2^N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

var result [][]int
func combinationSum(candidates []int, target int) [][]int {
    if len(candidates) == 0{
        return [][]int{}
    }
    
    result = [][]int{}
    
    helper(candidates,target,0,[]int{})
    
    return result
}


Without backtrack with extra space
func helper(candidates []int, target int,index int,path []int){
    if target == 0{
        result = append(result,path)
        return
    }
    if target < 0 || index ==  len(candidates){
        return
    }
       
    //0 case
    newpath1 := []int{}
    newpath1 = append(newpath1,path...)    
    helper(candidates,target,index+1,newpath1)
    
    //1 case
    newpath2 := []int{}
    newpath2 = append(newpath2,path...)
    newpath2 = append(newpath2,candidates[index])
    helper(candidates,target-candidates[index],index,newpath2)
    
}

//With backtrack and no extra space
func helper(candidates []int, target int,index int,path []int){
    if target == 0{
        newpath := []int{}
        newpath = append(newpath,path...)
        result = append(result,newpath)
        return
    }
    if target < 0 || index ==  len(candidates){
        return
    }
       
    //0 case    
    helper(candidates,target,index+1,path)
    
    //1 case
    path = append(path,candidates[index])
    helper(candidates,target-candidates[index],index,path)
    path = path[:len(path)-1]
}

File: Problem_1.js
// Combination Sum (https://leetcode.com/problems/combination-sum/)

// Time Complexity: Exponential
// Space Complexity: O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

let result;
var helper = (candidates, target, index, path) => {
    // Base
    if (target < 0) {
        return;
    } else if (target === 0) {
        result.push(Array.from(path));
        return;
    }
    // Logic
    for (let i = index; i < candidates.length; i++) {
        // Action
        path.push(candidates[i]);
        // Recurse
        helper(candidates, target - candidates[i], i, path);
        // Backtrack
        path.pop();
    }
}

/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
    if (candidates === null || candidates.length === 0)
        return [];
    result = [];
    helper(candidates, target, 0, []);
    return result;
};